### PR TITLE
Enable concatenation of unpadded prompts and a few other optimizations

### PR DESCRIPTION
**Pull request by sixiang-google**
# Description

-Based on https://github.com/AI-Hypercomputer/maxtext/pull/1162 and https://github.com/AI-Hypercomputer/maxtext/pull/1168, allow unpadded seqs to attend to each other.
-Enable overloading under len of 1024.
-Sorted data for less slot idle.
-Reconfigured garbage collection.

# Tests

Please describe how you tested this change, and include any instructions and/or
commands to reproduce.

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [X] I have performed a self-review of my code.
- [X] I have necessary comments in my code, particularly in hard-to-understand areas.
- [X] I have run end-to-end tests tests and provided workload links above if applicable.
- [X] I have made or will make corresponding changes to the doc if needed.

### CODE REVIEW

File: MaxText/inference_mlperf/offline_inference.py [hunk 1]
```diff
@@ -339,37 +358,57 @@ def detokenize():
...
-      # directly prefill prompts with 64 or less tokens, and with 1024 tokens
-      if num_tokens in (64, 1024) or not self.enable_batch_prefill:
-        first_token, slot, row = prefill([(slot, row)], num_tokens)[0]
+      # directly prefill prompts
+      if not self.enable_batch_prefill:
+        first_token, slot, row = prefill([(slot, row)], padded_len)[0]
+        self.detokenize_backlog.put((first_token, True, row.id, slot), block=True)
+        continue
+
+      if self.prefill_buckets[padded_len // 2] != []:
```
Comment (vipannalla) [inline MaxText/inference_mlperf/offline_inference.py:375]: nit: use len(self.prefill_buckets[padded_len // 2]) > 0.

File: MaxText/inference_mlperf/offline_inference.py [hunk 2]
```diff
@@ -379,6 +418,18 @@ def detokenize():
...
+    data_dict = defaultdict(list)
+    log.info("sorting data")
+    for row in data:
+      data_dict[row.tokens.shape[0]].append(row)
+    data_dict[128] += data_dict[64]
+    data_dict[64] = []
+    data = []
+    for padded_len in np.power([7, 8, 9, 10], 2):
+      log.info(f"padded len: {padded_len}, num: {len(data_dict[padded_len])}")
+      random.shuffle(data_dict[padded_len])
```
Comment (vipannalla) [inline MaxText/inference_mlperf/offline_inference.py:430]: Does random shuffle help with performance?

File: MaxText/inference_mlperf/offline_mode.py [hunk 3]
```diff
@@ -359,6 +358,7 @@ def flush_queries(self):
 
     log.info("Flush queries end")
     end = time.perf_counter()
+    gc.collect()
```
Comment (vipannalla) [inline MaxText/inference_mlperf/offline_mode.py:361]: What was the performance improvement with GC at the end vs. before each group?

**Review by vipannalla — COMMENTED** [review]
Looks good

**Review by vipannalla — APPROVED** [review]

File: MaxText/inference_mlperf/offline_inference.py [hunk 4]
```diff
@@ -105,32 +107,35 @@ def warmup(self, max_length, warmup_samples):
...
-              true_lengths=jnp.full(8, length, dtype=int),
-              decode_state=self.decode_state,
-          )
-          .compile()
-      )
+      min_num_prompts = max_length // length
+      max_num_prompts = max_length // (length // 2)
+      possible_prompts = range(min_num_prompts, max_num_prompts)
+      for num_prompts in possible_prompts:
+        log.info(f"Compiling batched prefill: {length} num_prompts: {num_prompts}")
```
[the remark this answers is not in the document]
Comment (sixiang-google) [reply MaxText/inference_mlperf/offline_inference.py:115]: Warmup now takes ~8min. Previously 2min

File: MaxText/inference_mlperf/offline_inference.py (hunk 2 above)
Comment (sixiang-google) [reply MaxText/inference_mlperf/offline_inference.py:430]: No but it makes the buckets spread out a little bit. Otherwise its possible that more than 15 prompts get in one prefill seq. The difference is small though.

File: MaxText/inference_mlperf/offline_mode.py (hunk 3 above)
Comment (sixiang-google) [reply MaxText/inference_mlperf/offline_mode.py:361]: about 2%. 4650->4750
